parser/src/string.rs

```diff
@@ -0,0 +1,189 @@
+                .map(|value| match value.node {
+                    ExprKind::JoinedStr { values } => Ok(values),
+                    ExprKind::Constant { .. } => Ok(vec![value]),
+                    _ => Err(LexicalError {
+                        location,
+                        error: LexicalErrorType::OtherError(
+                            "Unexpected expression kind in string concatenation.".to_string(),
+                        ),
+                    }),
+                })
+                .flat_map(|result| match result {
+                    Ok(values) => values.into_iter().map(Ok).collect(),
+                    Err(e) => vec![Err(e)],
+                })
+                .collect::<Result<Vec<Expr>, LexicalError>>()?;
+
+            // De-duplicate adjacent constants.
+            let mut deduped: Vec<Expr> = vec![];
+            let mut current: Vec<String> = vec![];
+            for value in values {
+                match value.node {
+                    ExprKind::FormattedValue { .. } => {
+                        if !current.is_empty() {
+                            deduped.push(Expr::new(
+                                location,
+                                ExprKind::Constant {
+                                    value: Constant::Str(current.join("")),
+                                    kind: kind.clone(),
+                                },
+                            ));
+                            current.clear();
+                        }
+                        deduped.push(value)
+                    }
+                    ExprKind::Constant { value, .. } => {
+                        if let Constant::Str(value) = value {
+                            current.push(value);
+                        } else {
+                            return Err(LexicalError {
+                                location,
+                                error: LexicalErrorType::OtherError(
+                                    "Unexpected non-string constant.".to_string(),
+                                ),
+                            });
+                        }
+                    }
+                    _ => {
+                        return Err(LexicalError {
+                            location,
+                            error: LexicalErrorType::OtherError(
+                                "Unexpected expression kind in string concatenation.".to_string(),
+                            ),
+                        });
+                    }
+                }
+            }
+
+            if !current.is_empty() {
+                deduped.push(Expr::new(
+                    location,
+                    ExprKind::Constant {
+                        value: Constant::Str(current.join("")),
+                        kind,
+                    },
+                ));
+                current.clear();
+            }
+
+            Expr::new(location, ExprKind::JoinedStr { values: deduped })
```

Comment:
The new JoinedStr creation will look like:
```suggestion
            if has_fstring {
                Expr::new(location, ExprKind::JoinedStr { values: deduped })
            } else {
                deduped.into_iter().exactly_one().expect("string must be concatenated to single element")
            }
```